packages/op-coreutils/src/op_coreutils/bigquery/write.py

```diff
@@ -1,150 +1,235 @@
+        table_name (str): The BigQuery table name.
+        expiration_days (int, optional): Partition expiration in days. Defaults to 360.
+    """
+    ensure_valid_dt(df, dt)
     df = df.with_columns(dt=pl.lit(dt).str.strptime(pl.Datetime, "%Y-%m-%d"))
     overwrite_partitions(df, dataset, table_name, expiration_days)
 
 
 def overwrite_partitions(
-    df: pl.DataFrame,
-    dataset: str,
-    table_name: str,
-    expiration_days: int = 360,
+    df: pl.DataFrame, dataset: str, table_name: str, expiration_days: int = 360
 ):
+    """Overwrite multiple partitions in a BigQuery table.
+
+    Args:
+        df (pl.DataFrame): The DataFrame to write.
+        dataset (str): The BigQuery dataset name.
+        table_name (str): The BigQuery table name.
+        expiration_days (int, optional): Partition expiration in days. Defaults to 360.
+    """
     client = init_client()
+    ensure_valid_dt(df)
 
     destination = f"{dataset}.{table_name}"
 
     if df["dt"].dtype == pl.String:
         df = df.with_columns(dt=pl.col("dt").str.strptime(pl.Datetime, "%Y-%m-%d"))
 
     partitions = df["dt"].unique().sort().to_list()
-    if isinstance(client, MagicMock):
-        operation = "DRYRUN "
-    else:
-        operation = ""
     log.info(
-        f"{operation}Writing {len(partitions)} partitions to BQ [{partitions[0]} ... {partitions[-1]}]"
+        f"Writing {len(partitions)} partitions to BQ [{partitions[0]} ... {partitions[-1]}]"
     )
 
-    with io.BytesIO() as stream:
-        df.write_parquet(stream)
-        filesize = stream.tell()
-        stream.seek(0)
-        job = client.load_table_from_file(
-            stream,
-            destination=destination,
-            job_config=bigquery.LoadJobConfig(
-                source_format=bigquery.SourceFormat.PARQUET,
-                write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
-                time_partitioning=bigquery.TimePartitioning(
-                    type_=bigquery.TimePartitioningType.DAY,
-                    field="dt",  # Name of the column to use for partitioning.
-                    expiration_ms=expiration_days * 24 * 3600 * 1000,
-                ),
-            ),
-        )
-        job.result()
+    time_partitioning = bigquery.TimePartitioning(
+        type_=bigquery.TimePartitioningType.DAY,
+        field="dt",
+        expiration_ms=expiration_days * 24 * 3600 * 1000,
+    )
 
-        if isinstance(client, MagicMock):
-            operation = "DRYRUN "
-        else:
-            operation = ""
+    write_df_to_bq(
+        df,
+        destination,
+        client,
+        operation="OVERWRITE PARTITION",
+        time_partitioning=time_partitioning,
+    )
 
-        log.info(
-            f"{operation}OVERWRITE PARTITION: Wrote {human_rows(len(df))} {human_size(filesize)} to BQ {destination}"
-        )
+
+def upsert_partition(
+    df: pl.DataFrame, dt: str, dataset: str, table_name: str, unique_keys=["dt"]
```

Comment:
so the default case is equivalent to the dt partition, but we can specify different partitions if/when needed?

Reply:
Correct, you can specify two keys for example